# Problem 1: Mean Molecular Weights

1) Calculate the values of the mean molecular weights under the following circumstances:

(a) all Hydrogen, i.e., $X = 1$ and $Y = Z = 0$; **both** (i) completely neutral and (ii) completely ionized gas;

(b) all Helium, i.e., $X = Z = 0$ and $Y = 1$; **both** (i) completely neutral and (ii) completely ionized gas;

(c) all heavy elements, i.e., $X = Y = 0$ and $Z = 1$; **both** (i) completely neutral and (ii) completely ionized gas;

(d) solar abundance, i.e., $X = 0.7$, $Y = 0.28$, and $Z = 0.02$; **both** (i) completely neutral and (ii) completely ionized gas.

**NOTE:** $\langle n_Z / A_Z \rangle$ is roughly equal to  
(i) $\sim 1/15.5$ for the **neutral** case at solar metallicity, and  
(ii) $(z + 1)/(2z + 2) \approx 1/2$ for the **fully ionized** case.

We assume $m_{\rm u}$ (the atomic mass unit) as the reference mass.


## Derivation: Mean Molecular Weight $\mu$

We define the **mean molecular weight** $\mu$ as  

$$
\mu \equiv \frac{\rho}{n \, m_{\rm u}},
$$

where  
- $\rho$ is the mass density,  
- $n$ is the **total number density of particles** (nuclei $+$ electrons, depending on ionization),  
- $m_{\rm u}$ is the atomic mass unit.

Equivalently,  

$$
\frac{1}{\mu} = \frac{n \, m_{\rm u}}{\rho}.
$$


### 1. Mass fractions

Let  
- $X$ = mass fraction of Hydrogen (H, $A=1$, $Z_{\rm nuc}=1$),  
- $Y$ = mass fraction of Helium (He, $A=4$, $Z_{\rm nuc}=2$),  
- $Z$ = mass fraction of "metals" (all heavier elements).  

Total mass density can be written as  

$$
\rho = \rho X + \rho Y + \rho Z = \rho (X + Y + Z),
$$

and by definition $X + Y + Z = 1$.

It is convenient to work **per unit mass** of gas. Take $\rho = 1$ for the derivation; then we can interpret number densities directly as “per gram” or “per $m_{\rm u}$”.


### 2. Neutral gas

For a **completely neutral** gas:
- each H atom contributes **1 particle** with mass $1 \times m_{\rm u}$,  
- each He atom contributes **1 particle** with mass $4 \times m_{\rm u}$,  
- each metal ion (neutral) contributes **1 particle** with mass $A_Z \times m_{\rm u}$.

Number density per unit mass:

$$
n_{\rm H} = \frac{X}{1 \cdot m_{\rm u}} = \frac{X}{m_{\rm u}}
$$

$$
n_{\rm He} = \frac{Y}{4 \cdot m_{\rm u}} = \frac{Y}{4 m_{\rm u}}
$$

Metals (neutral): the problem tells us to use  

$$
\left\langle \frac{n_Z}{A_Z} \right\rangle \approx \frac{1}{15.5}
$$

for solar metallicity. That means, for mass fraction $Z$, the number of neutral metal particles is approximately  

$$
n_Z \approx \frac{Z}{15.5 \, m_{\rm u}}.
$$

So the **total number density** (neutral case) is  

$$
n_{\rm neutral} = \frac{X}{m_{\rm u}} + \frac{Y}{4 m_{\rm u}} + \frac{Z}{15.5\, m_{\rm u}}.
$$

Then  

$$
\frac{1}{\mu_{\rm neutral}} = \frac{n_{\rm neutral} \, m_{\rm u}}{\rho}.
$$

If we are working per unit mass ($\rho = 1$), then  

$$
\frac{1}{\mu_{\rm neutral}} = X + \frac{Y}{4} + \frac{Z}{15.5}.
$$

Therefore  

$$
\boxed{ \mu_{\rm neutral} = \frac{1}{X + \dfrac{Y}{4} + \dfrac{Z}{15.5}} }.
$$


### 3. Fully ionized gas

For a **completely ionized** gas we must count **nuclei + electrons**.

- Hydrogen: each H atom $\to$ 1 proton + 1 electron = **2 particles** from mass $1 \, m_{\rm u}$  

$$
n_{\rm H,ion} = \frac{X}{m_{\rm u}} \times 2
\quad \Rightarrow \quad \text{contribution to } \frac{1}{\mu} = 2X.
$$

- Helium: each He atom (mass $4 m_{\rm u}$) $\to$ 1 He nucleus + 2 electrons = **3 particles**

$$
n_{\rm He,ion} = \frac{Y}{4 m_{\rm u}} \times 3
\quad \Rightarrow \quad \text{contribution to } \frac{1}{\mu} = \frac{3}{4}Y.
$$

- Metals: for the ionized case the problem gives  

$$
\frac{Z+1}{2Z+2} \simeq \frac{1}{2},
$$

interpreted as  

$$
\left\langle \frac{n_{Z,\text{ion}}}{1} \right\rangle \approx \frac{Z}{2 m_{\rm u}}.
$$

Thus metals contribute $\dfrac{Z}{2}$ to $1/\mu$.

Hence for a fully ionized mixture:  

$$
\boxed{
\frac{1}{\mu_{\rm ion}} = 2X + \frac{3}{4}Y + \frac{1}{2}Z
}
$$

and therefore  

$$
\boxed{
\mu_{\rm ion} = \frac{1}{2X + \dfrac{3}{4}Y + \dfrac{1}{2}Z}.
}
$$


### 4. Now plug in special cases

1. **All H:** $X=1$, $Y=Z=0$
   $$
   \mu_{\text{neutral}} = 1, \quad \mu_{\text{ionized}} = 0.5
   $$

2. **All He:** $X=0$, $Y=1$
   $$
   \mu_{\text{neutral}} = 4, \quad \mu_{\text{ionized}} = \frac{4}{3} \approx 1.33
   $$

3. **All metals:** $X=Y=0$, $Z=1$
   $$
   \mu_{\text{neutral}} = 15.5, \quad \mu_{\text{ionized}} = 2
   $$

4. **Solar mixture:** $X=0.7$, $Y=0.28$, $Z=0.02$
   $$
   \frac{1}{\mu_{\odot,\text{neutral}}}
   = 0.7 + \frac{0.28}{4} + \frac{0.02}{15.5}
   $$
   $$
   \frac{1}{\mu_{\odot,\text{ion}}}
   = 2(0.7) + \frac{3}{4}(0.28) + \frac{1}{2}(0.02)
   $$
   These will be evaluated numerically in the next code cell.


In [4]:
# Mean molecular weight calculator for different compositions

def mu_neutral(X, Y, Z):
    """
    Mean molecular weight for a neutral gas.
    Uses:
      1/mu = X + Y/4 + Z/15.5
    per problem statement.
    """
    return 1.0 / (X + Y/4.0 + Z/15.5)

def mu_ionized(X, Y, Z):
    """
    Mean molecular weight for a fully ionized gas.
    Uses:
      1/mu = 2X + 3/4 Y + 1/2 Z
    per problem statement.
    """
    return 1.0 / (2.0*X + 0.75*Y + 0.5*Z)


# (a) all Hydrogen
mu_H_neutral = mu_neutral(1.0, 0.0, 0.0)
mu_H_ion     = mu_ionized(1.0, 0.0, 0.0)

# (b) all Helium
mu_He_neutral = mu_neutral(0.0, 1.0, 0.0)
mu_He_ion     = mu_ionized(0.0, 1.0, 0.0)

# (c) all metals
mu_Z_neutral = mu_neutral(0.0, 0.0, 1.0)
mu_Z_ion     = mu_ionized(0.0, 0.0, 1.0)

# (d) solar
X_solar, Y_solar, Z_solar = 0.7, 0.28, 0.02
mu_solar_neutral = mu_neutral(X_solar, Y_solar, Z_solar)
mu_solar_ion     = mu_ionized(X_solar, Y_solar, Z_solar)

# Show results
results = {
    "H (neutral)": mu_H_neutral,
    "H (ionized)": mu_H_ion,
    "He (neutral)": mu_He_neutral,
    "He (ionized)": mu_He_ion,
    "metals (neutral)": mu_Z_neutral,
    "metals (ionized)": mu_Z_ion,
    "solar (neutral)": mu_solar_neutral,
    "solar (ionized)": mu_solar_ion,
}

for k, v in results.items():
    print(f"{k:20s} : {v:.6f}")


H (neutral)          : 1.000000
H (ionized)          : 0.500000
He (neutral)         : 4.000000
He (ionized)         : 1.333333
metals (neutral)     : 15.500000
metals (ionized)     : 2.000000
solar (neutral)      : 1.296529
solar (ionized)      : 0.617284


## Final Results

We summarize the mean molecular weights $\mu$ for each case.

**(a) All Hydrogen, $X=1$, $Y=Z=0$**

$$
\boxed{ \mu_{\text{neutral}} = 1.0 }
$$

$$
\boxed{ \mu_{\text{ionized}} = 0.5 }
$$


**(b) All Helium, $Y=1$**

$$
\boxed{ \mu_{\text{neutral}} = 4.0 }
$$

$$
\boxed{ \mu_{\text{ionized}} = \frac{4}{3} \approx 1.333 }
$$

**(c) All heavy elements, $Z=1$**

$$
\boxed{ \mu_{\text{neutral}} = 15.5 }
$$

$$
\boxed{ \mu_{\text{ionized}} = 2.0 }
$$


**(d) Solar mixture, $X=0.7$, $Y=0.28$, $Z=0.02$**

First, neutral:

$$
\frac{1}{\mu_{\odot,\,\text{neutral}}}
= 0.7 + \frac{0.28}{4} + \frac{0.02}{15.5}
= 0.7 + 0.07 + 0.00129 \approx 0.77129,
$$

so

$$
\boxed{ \mu_{\odot,\,\text{neutral}} \approx \frac{1}{0.77129} \approx 1.30 }.
$$

Then, fully ionized:

$$
\frac{1}{\mu_{\odot,\,\text{ion}}}
= 2(0.7) + \frac{3}{4}(0.28) + \frac{1}{2}(0.02)
= 1.4 + 0.21 + 0.01
= 1.62,
$$

so

$$
\boxed{ \mu_{\odot,\,\text{ion}} = \frac{1}{1.62} \approx 0.617 }.
$$


## Discussion

1. The pattern is exactly what we expect physically:  
   - **Neutral gas** has **fewer particles per unit mass** (because there are no free electrons), so $\mu$ is **larger**.  
   - **Fully ionized gas** has **more particles per unit mass** (nuclei + electrons), so $\mu$ is **smaller**.


2. Pure hydrogen is the most transparent case:  
   - neutral H: 1 proton + 1 electron **bound** $\Rightarrow$ 1 particle $\Rightarrow \mu = 1$,  
   - ionized H: 1 proton + 1 free electron $\Rightarrow$ 2 particles $\Rightarrow \mu = 0.5$.  

   This is the canonical “fully ionized hydrogen plasma” value that shows up in stellar structure.


3. Pure helium jumps to $\mu = 4$ in the neutral case because you get only **1 particle per 4 amu**.  
   When ionized you get 3 particles per 4 amu $\Rightarrow \mu = 4/3$.


4. Metals look “heavy” in the neutral case because the problem told us to use  
   $\langle n_Z/A_Z \rangle \approx 1/15.5$,  
   so you get **very few particles per unit mass** $\Rightarrow$ very large $\mu \approx 15.5$.  
   Once ionized, their contribution per unit mass is taken to be $\sim 1/2$,  
   so you get $\mu \approx 2$.


5. The **solar mixture** values:

$$
\mu_{\odot,\,\text{neutral}} \approx 1.3, \quad
\mu_{\odot,\,\text{ion}} \approx 0.62
$$

are exactly in the ballpark of what we use in stellar and galactic astrophysics.  
The ionized value $\sim 0.6$ is the standard number used for fully ionized, near-solar-composition gas  
(e.g. in hot stars or intracluster medium approximations).



## Problem 2: Mean Molecular Weight per Electron, $\mu_e$

**2)** Calculate the **mean molecular weight per electron**, $\mu_e$, for **completely ionized** conditions of  
(a) all Hydrogen ($X = 1$ and $Y = Z = 0$);  
(b) all Helium ($X = Z = 0$, $Y = 1$); and  
(c) all heavy elements, i.e. $X = Y = 0$ and $Z = 1$.

**NOTE:** $\langle n_{Z-1} / A_Z \rangle \approx 0.5$ for the fully ionized case.  


## Derivation: Mean Molecular Weight per Electron $\mu_e$

We now want **$\mu_e$**, not the total mean molecular weight $\mu$.

By definition,
$$
\mu_e \equiv \frac{\rho}{n_e \, m_{\rm u}},
$$
where
- $\rho$ is the mass density,
- $n_e$ is the **number density of free electrons**,
- $m_{\rm u}$ is the atomic mass unit.

Equivalently,
$$
\frac{1}{\mu_e} = \frac{n_e \, m_{\rm u}}{\rho}.
$$

We are told to assume **completely ionized** gas. That means:
- every H atom contributes **1 free electron**,
- every He atom contributes **2 free electrons**,
- every heavy element (metal) contributes on average **0.5 electrons per baryon** (as told to us by: $\langle n_{Z-1} / A_Z \rangle \approx 0.5$).

We will again work **per unit mass** (set $\rho = 1$) so that counting particles per $m_{\rm u}$ is easy.


### General setup

Let $X, Y, Z$ be the mass fractions of H, He, and metals, respectively, with
$$
X + Y + Z = 1.
$$

For a **completely ionized** mixture:

1. **Hydrogen** ($A=1$, 1 electron when ionized):  
   mass fraction $X$ gives
   $$
   n_{e,{\rm H}} = \frac{X}{1 \cdot m_{\rm u}} \times 1 = \frac{X}{m_{\rm u}}.
   $$

2. **Helium** ($A=4$, 2 electrons when ionized):  
   mass fraction $Y$ gives
   $$
   n_{e,{\rm He}} = \frac{Y}{4 m_{\rm u}} \times 2 = \frac{Y}{2 m_{\rm u}}.
   $$

3. **Metals** (fully ionized): the problem tells us to use
   $$
   \left\langle \frac{n_{Z-1}}{A_Z} \right\rangle \approx 0.5.
   $$
   This we interpret as: **for mass fraction $Z$ of heavy elements, the number of electrons per unit mass is**
   $$
   n_{e,{\rm Z}} \approx \frac{Z}{m_{\rm u}} \times 0.5 = \frac{0.5\, Z}{m_{\rm u}}.
   $$

So the **total** electron number density is
$$
n_e = \frac{X}{m_{\rm u}} + \frac{Y}{2 m_{\rm u}} + \frac{0.5 Z}{m_{\rm u}}.
$$

Factor out $1/m_{\rm u}$:
$$
n_e = \frac{1}{m_{\rm u}} \left( X + \frac{Y}{2} + \frac{Z}{2} \right).
$$

Now plug into the definition
$$
\frac{1}{\mu_e} = \frac{n_e \, m_{\rm u}}{\rho}.
$$

If we work per unit mass ($\rho = 1$), then
$$
\frac{1}{\mu_e} = X + \frac{Y}{2} + \frac{Z}{2}.
$$

Therefore the **general formula** (for fully ionized gas) is
$$
\boxed{
\mu_e = \frac{1}{X + \dfrac{Y}{2} + \dfrac{Z}{2}}
}
$$


### Now evaluate the three special cases

#### (a) All Hydrogen
Here $X = 1$, $Y = 0$, $Z = 0$.
Then
$$
\frac{1}{\mu_e} = 1 \quad \Rightarrow \quad \boxed{\mu_{e,{\rm H}} = 1.0}.
$$
This makes sense: 1 proton $\to$ 1 electron.


#### (b) All Helium
Here $X=0$, $Y=1$, $Z=0$.
Then
$$
\frac{1}{\mu_e} = \frac{1}{2} \quad \Rightarrow \quad \boxed{\mu_{e,{\rm He}} = 2.0}.
$$
This also makes sense: fully ionized He has 2 electrons for 4 amu $\Rightarrow$ 1 electron per 2 amu.


#### (c) All heavy elements
Here $X=0$, $Y=0$, $Z=1$.
The problem **tells** us to use $ \langle n_{Z-1}/A_Z \rangle \approx 0.5 $, so
$$
\frac{1}{\mu_e} = \frac{1}{2} \quad \Rightarrow \quad \boxed{\mu_{e,{\rm metals}} = 2.0}.
$$

So the answers are:
- H (fully ionized): $\mu_e = 1$,
- He (fully ionized): $\mu_e = 2$,
- Metals (fully ionized, using the note): $\mu_e = 2$.


## Discussion

1. This problem is *simpler* than the total mean molecular weight problem because we only count **electrons**, not nuclei $+$ electrons.

2. For a **fully ionized** gas, $\mu_e$ is controlled by “how many electrons you get per unit mass.”
   - Hydrogen gives **1 electron per 1 amu** $\Rightarrow$ $\mu_e = 1$.
   - Helium gives **2 electrons per 4 amu** $\Rightarrow$ $1$ electron per $2$ amu $\Rightarrow$ $\mu_e = 2$.

3. For **metals**, the problem statement gives us the average ratio
   $$
   \left\langle \frac{n_{Z-1}}{A_Z} \right\rangle \approx 0.5,
   $$
   which is basically saying: “pretend metals give you **half an electron per amu** on average when fully ionized.”  
   That is the **same electron-per-mass** ratio as ionized helium, so we get the **same** $\mu_e = 2$.

4. These values are standard benchmarks in stellar/galactic astrophysics:
   - H plasma: $\mu_e = 1$
   - solar-like / He-like electron-per-baryon: $\mu_e \sim 2$
   Realistic mixtures (like $X=0.7, Y=0.28, Z=0.02$) would land **between** 1 and 2, typically around $\mu_e \sim 1.15$–$1.2$ depending on exact treatment of metals.

5. The important pattern to remember:
   $$
   \text{(electrons per unit mass)} \quad \Longrightarrow \quad \frac{1}{\mu_e} = X + \frac{Y}{2} + \frac{Z}{2}.
   $$
   Then just plug in the composition.


## Problem 3: Equation of State with Gas + Radiation Pressure

The center of a certain star contains 70% Hydrogen and 28% Helium by weight (i.e., $X=0.7$, $Y=0.28$, $Z=0.02$). **Evaluate numerically the equation of state** (including both gas pressure and radiation pressure).

**Given:** $\rho = 50~\mathrm{g~cm^{-3}}$, $T = 15\times 10^6~\mathrm{K}$.

**Tasks:**
1. Compute $P_{\rm gas}$ and $P_{\rm rad}$.
2. Compare them by reporting $P_{\rm gas}/P_{\rm rad}$.
3. Also report $\beta \equiv P_{\rm gas}/(P_{\rm gas}+P_{\rm rad})$ and the ratio $(1-\beta)/\beta$.

**Note:** In cgs, $1~\mathrm{erg\,cm^{-3}} = 1~\mathrm{dyne\,cm^{-2}}$.


## Derivation

We include **ideal gas** + **radiation**:

$$
P = P_{\rm gas} + P_{\rm rad}.
$$

### Gas pressure
For a fully ionized mixture,
$$
P_{\rm gas} = \frac{\rho k_B T}{\mu m_{\rm u}},
$$
where $k_B$ is Boltzmann's constant, $m_{\rm u}$ is the atomic mass unit, and $\mu$ is the **mean molecular weight** of the ions + electrons mixture.

For a fully ionized gas with mass fractions $(X,Y,Z)$,
$$
\frac{1}{\mu} = 2X + \frac{3}{4}Y + \frac{1}{2}Z.
$$
For $(X,Y,Z)=(0.7,\,0.28,\,0.02)$,
$$
\frac{1}{\mu} = 2(0.7) + \frac{3}{4}(0.28) + \frac{1}{2}(0.02)
= 1.62 \;\;\Rightarrow\;\; \mu = \frac{1}{1.62} \approx 0.61728395.
$$

Hence
$$
P_{\rm gas} = \frac{\rho k_B T}{\mu m_{\rm u}}.
$$

### Radiation pressure
The radiation pressure is
$$
P_{\rm rad} = \frac{1}{3} a T^4,
$$
with $a = 7.5657\times 10^{-15}\ \mathrm{erg\;cm^{-3}\;K^{-4}}$.

### Ratios
Define
$$
\beta \equiv \frac{P_{\rm gas}}{P_{\rm gas}+P_{\rm rad}},
\qquad
\frac{P_{\rm rad}}{P_{\rm gas}} = \frac{1-\beta}{\beta},
\qquad
\frac{P_{\rm gas}}{P_{\rm rad}} = \frac{\beta}{1-\beta}.
$$

We will compute $P_{\rm gas}$, $P_{\rm rad}$, then $\beta$ and the ratios.


In [6]:
# cgs constants
k_B  = 1.380649e-16       # erg K^-1
m_u  = 1.66053906660e-24  # g
a    = 7.5657e-15         # erg cm^-3 K^-4

# inputs
rho = 50.0          # g cm^-3
T   = 15.0e6        # K
X, Y, Z = 0.7, 0.28, 0.02

# mean molecular weight (fully ionized mixture)
mu = 1.0 / (2.0*X + 0.75*Y + 0.5*Z)

# pressures
P_gas = rho * k_B * T / (mu * m_u)   # erg cm^-3 = dyne cm^-2
P_rad = (1.0/3.0) * a * T**4         # erg cm^-3 = dyne cm^-2
P_tot = P_gas + P_rad

# beta and ratios
beta = P_gas / P_tot
Prad_over_Pgas = (1.0 - beta) / beta
Pgas_over_Prad = beta / (1.0 - beta)

print(f"mu (fully ionized)     = {mu:.9f}")
print(f"P_gas [dyne/cm^2]      = {P_gas:.6e}")
print(f"P_rad [dyne/cm^2]      = {P_rad:.6e}")
print(f"P_tot [dyne/cm^2]      = {P_tot:.6e}")
print(f"beta = P_gas/P_tot     = {beta:.9f}")
print(f"P_rad/P_gas            = {Prad_over_Pgas:.6e}")
print(f"P_gas/P_rad            = {Pgas_over_Prad:.6e}")


mu (fully ionized)     = 0.617283951
P_gas [dyne/cm^2]      = 1.010207e+17
P_rad [dyne/cm^2]      = 1.276712e+14
P_tot [dyne/cm^2]      = 1.011484e+17
beta = P_gas/P_tot     = 0.998737783
P_rad/P_gas            = 1.263812e-03
P_gas/P_rad            = 7.912570e+02


## Final Results (boxed)

With $\rho = 50~\mathrm{g\,cm^{-3}}$, $T = 1.5\times 10^7~\mathrm{K}$, and $(X,Y,Z)=(0.7,0.28,0.02)$:

$$
\boxed{\mu \approx 0.61728395}
$$

Gas pressure:
$$
\boxed{P_{\rm gas} \approx 1.0102\times 10^{17}~\mathrm{dyne\,cm^{-2}}}
$$

Radiation pressure:
$$
\boxed{P_{\rm rad} \approx 1.2767\times 10^{14}~\mathrm{dyne\,cm^{-2}}}
$$

Total:
$$
\boxed{P_{\rm tot} \approx 1.0115\times 10^{17}~\mathrm{dyne\,cm^{-2}}}
$$

Gas dominance parameter:
$$
\boxed{\beta \equiv \frac{P_{\rm gas}}{P_{\rm gas}+P_{\rm rad}} \approx 0.998738}
$$

Ratios:
$$
\boxed{\frac{P_{\rm gas}}{P_{\rm rad}} \approx 7.91\times 10^{2}}
\qquad
\boxed{\frac{P_{\rm rad}}{P_{\rm gas}} \approx 1.26\times 10^{-3}}
$$

(So radiation pressure is $\sim 0.13\%$ of the gas pressure.)


## Discussion

1. **Gas vs radiation:** Gas pressure overwhelmingly dominates at these conditions:
   $P_{\rm gas}/P_{\rm rad} \approx 7.9\times 10^2$. Radiation pressure is only
   $\sim 0.13\%$ of the total.

2. **Why:** Even though $T=1.5\times 10^7~\mathrm{K}$ is high, the density is modest for a
   stellar center and the composition yields a small $\mu\approx 0.617$, which boosts
   $P_{\rm gas}=\rho k_B T/(\mu m_{\rm u})$. Radiation pressure scales as $T^4$ and is
   important only at **very** high $T$ (massive star cores, very hot envelopes, etc.).

3. **On the $\beta$ notation:** By definition $\beta=P_{\rm gas}/(P_{\rm gas}+P_{\rm rad})$,
   so
   $$
   \frac{P_{\rm rad}}{P_{\rm gas}}=\frac{1-\beta}{\beta},\qquad
   \frac{P_{\rm gas}}{P_{\rm rad}}=\frac{\beta}{1-\beta}.
   $$
   Plugging our $\beta\simeq 0.998738$ gives
   $P_{\rm rad}/P_{\rm gas}\simeq 1.26\times 10^{-3}$ and
   $P_{\rm gas}/P_{\rm rad}\simeq 7.91\times 10^{2}$, in agreement with the direct
   pressure calculation.

4. **Units:** All pressures are reported in cgs; $1~\mathrm{erg\,cm^{-3}} = 1~\mathrm{dyne\,cm^{-2}}$.

## Problem 4: Non-relativistic limit of the completely degenerate electron pressure

Consider the pressure of completely degenerate, relativistic electrons
$$
P_{e,r}=\frac{\pi m_0^4 c^5}{3h^3}\,f(x), \qquad x \equiv \frac{p_0}{m_0 c},
$$
with
$$
f(x)=x(2x^2-3)\sqrt{1+x^2}+3\sinh^{-1}x.
$$

**Task.** Show that for $x\to 0$ (i.e. $p_0\ll m_0 c$),
$$
f(x)\approx \frac{8}{5}x^5-\frac{4}{7}x^7+\cdots,
$$
and confirm that this gives the completely degenerate **non-relativistic** electron pressure
$$
P_{e,{\rm nr}}=\frac{8\pi}{15\,m\,h^3}\,p_0^5.
$$

**Tip:** Taylor expand both $\sqrt{1+x^2}$ and $\sinh^{-1}x$ to four terms for $x\to 0$.


## Derivation

We need the $x\to 0$ expansions. Using standard series:

1) Binomial (with $y=x^2$):

$$
\sqrt{1+x^2}=(1+y)^{1/2}
=1+\frac{1}{2}y-\frac{1}{8}y^2+\frac{1}{16}y^3-\cdots
=1+\frac{x^2}{2}-\frac{x^4}{8}+\frac{x^6}{16}+\mathcal{O}(x^8).
$$

2) Inverse hyperbolic sine:

$$
\sinh^{-1}x
= x-\frac{x^3}{6}+\frac{3x^5}{40}-\frac{5x^7}{112}+\mathcal{O}(x^9).
$$

Now expand

$$
f(x)=x(2x^2-3)\sqrt{1+x^2}+3\,\sinh^{-1}x.
$$

### First term:

$$
x(2x^2-3)\sqrt{1+x^2}
= x(2x^2-3)\left(1+\frac{x^2}{2}-\frac{x^4}{8}+\frac{x^6}{16}\right)+\mathcal{O}(x^9).
$$

Distribute $x(2x^2-3)=2x^3-3x$ and multiply term-by-term (keep through $x^7$):

- With $1$: $(2x^3-3x)$  
- With $x^2/2$: $(2x^3-3x)\cdot \frac{x^2}{2} = x^5-\frac{3}{2}x^3$  
- With $-x^4/8$: $(2x^3-3x)\cdot\left(-\frac{x^4}{8}\right)= -\frac{1}{4}x^7+\frac{3}{8}x^5$  
- With $x^6/16$: $(2x^3-3x)\cdot\frac{x^6}{16}= \frac{1}{8}x^9-\frac{3}{16}x^7 \ \ (\text{drop }x^9)$

Collect like powers up to $x^7$:

$$
\begin{aligned}
x(2x^2-3)\sqrt{1+x^2}
&= (-3x) + \left(2x^3 - \frac{3}{2}x^3\right)
+ \left(x^5+\frac{3}{8}x^5\right)
+ \left(-\frac{1}{4}x^7 - \frac{3}{16}x^7\right)
+ \mathcal{O}(x^9) \\
&= -3x + \frac{1}{2}x^3 + \frac{11}{8}x^5 - \frac{7}{16}x^7 + \mathcal{O}(x^9).
\end{aligned}
$$

### Second term:

$$
3\,\sinh^{-1}x
= 3x - \frac{1}{2}x^3 + \frac{9}{40}x^5 - \frac{15}{112}x^7 + \mathcal{O}(x^9).
$$

### Sum:

$$
\begin{aligned}
f(x) &= \left(-3x + \frac{1}{2}x^3 + \frac{11}{8}x^5 - \frac{7}{16}x^7\right)
+ \left(3x - \frac{1}{2}x^3 + \frac{9}{40}x^5 - \frac{15}{112}x^7\right)
+ \mathcal{O}(x^9) \\
&= \left(\cancel{-3x}+\cancel{3x}\right)
+ \left(\tfrac{1}{2}x^3 - \tfrac{1}{2}x^3\right)
+ \left(\frac{11}{8}+\frac{9}{40}\right)x^5
+ \left(-\frac{7}{16} - \frac{15}{112}\right)x^7
+ \mathcal{O}(x^9) \\
&= \left(\frac{55}{40}+\frac{9}{40}\right)x^5
+ \left(-\frac{49}{112} - \frac{15}{112}\right)x^7
+ \mathcal{O}(x^9) \\
&= \frac{64}{40}x^5 - \frac{64}{112}x^7 + \mathcal{O}(x^9) \\
&= \boxed{\frac{8}{5}x^5 - \frac{4}{7}x^7 + \cdots }.
\end{aligned}
$$

### Plug into $P_{e,r}$

With $x = p_0/(m_0 c)$,

$$
P_{e,r}=\frac{\pi m_0^4 c^5}{3h^3}
\left(\frac{8}{5}\frac{p_0^5}{m_0^5 c^5} - \frac{4}{7}\frac{p_0^7}{m_0^7 c^7} + \cdots\right)
= \underbrace{\frac{8\pi}{15\,m_0 h^3}p_0^5}_{\displaystyle P_{e,{\rm nr}}}
\ -\ \frac{4\pi}{21\,m_0^3 c^2 h^3}p_0^7 + \cdots
$$

Hence the **leading term** recovers the completely degenerate **non-relativistic** result:

$$
\boxed{P_{e,{\rm nr}}=\frac{8\pi}{15\,m\,h^3}\,p_0^5.}
$$


In [8]:
import math

def f_exact(x):
    return x*(2*x**2 - 3)*math.sqrt(1 + x**2) + 3*math.asinh(x)

def f_series(x):
    # 8/5 x^5 − 4/7 x^7
    return (8/5)*x**5 - (4/7)*x**7

# Show the approximate Taylor expansion coefficients manually
print("Expected series: f(x) ≈ (8/5)x^5 − (4/7)x^7")

# Compare values
print("\nCompare exact vs truncated series:")
print(f"{'x':>10} {'f_exact':>20} {'f_series':>20} {'rel.err':>12}")
for val in [1e-3, 1e-2, 5e-2, 1e-1]:
    fe = f_exact(val)
    fs = f_series(val)
    relerr = abs((fe - fs) / fe) if fe != 0 else 0.0
    print(f"{val:10.3e} {fe:20.10e} {fs:20.10e} {relerr:12.3e}")



Expected series: f(x) ≈ (8/5)x^5 − (4/7)x^7

Compare exact vs truncated series:
         x              f_exact             f_series      rel.err
 1.000e-03     1.5994150449e-15     1.5999994286e-15    3.654e-04
 1.000e-02     1.5999428518e-10     1.5999428571e-10    3.330e-09
 5.000e-02     4.9955422135e-07     4.9955357143e-07    1.301e-06
 1.000e-01     1.5943188220e-05     1.5942857143e-05    2.077e-05


## Final Results (boxed)

Small-$x$ expansion:
$$
\boxed{\,f(x)\approx \frac{8}{5}x^5 - \frac{4}{7}x^7 + \cdots\,}
$$

Non-relativistic degenerate pressure recovered:
$$
\boxed{\,P_{e,{\rm nr}}=\frac{8\pi}{15\,m\,h^3}\,p_0^5\,}
$$

with the next correction (relativistic) term:
$$
\boxed{\,P_{e,r}= \frac{8\pi}{15\,m h^3}p_0^5
\ -\ \frac{4\pi}{21\,m^3 c^2 h^3}p_0^7 + \cdots\,}
$$


## Discussion

- The apparent $\mathcal{O}(x)$ and $\mathcal{O}(x^3)$ terms **cancel exactly** between $x(2x^2-3)\sqrt{1+x^2}$ and $3\sinh^{-1}x$, leaving $x^5$ as the leading contribution. This explains why the non-relativistic limit scales as $p_0^5$.

- Substituting $x = p_0/(m c)$ into $P_{e,r} = (\pi m^4 c^5 / 3h^3)\,f(x)$ gives $P \propto p_0^5$ at leading order, matching the standard completely degenerate, non-relativistic expression derived in class.

- The next term, $-\,(4\pi/21)\,p_0^7/(m^3 c^2 h^3)$, is the first relativistic correction and becomes significant only when $p_0 \sim m c$.

- **Takeaway:** the relativistic expression reproduces the non-relativistic limit smoothly, and the series expansion provides controlled way to estimate small relativistic corrections when $x \ll 1$.


## Problem 5: Ultra-relativistic (large-x) limit of degenerate electron pressure

From Problem 4, the pressure of completely degenerate, relativistic electrons is
$$
P_{e,r}=\frac{\pi m_0^4 c^5}{3h^3}\,f(x), \qquad x \equiv \frac{p_0}{m_0 c},
$$
with
$$
f(x)=x(2x^2-3)\sqrt{1+x^2}+3\sinh^{-1}x.
$$

**Task.** Show that for $x\to\infty$ (highly relativistic degeneracy),
$$
f(x)\approx 2x^4-2x^2+\cdots,
$$
and confirm that inserting this into $P_{e,r}$ yields the same ultra-relativistic pressure as obtained by setting $v_p=c$ in the pressure integral.

**Tip:** Taylor expand (3 terms) for $x\to\infty$:
$$
\sqrt{1+x^2}=x\left(1+\frac{1}{x^2}\right)^{1/2},\qquad
\sinh^{-1}x=\ln\!\Big[x+\sqrt{1+x^2}\,\Big].
$$


## Derivation

We need large-$x$ expansions. Write

$$
\sqrt{1+x^2}=x\sqrt{1+\frac{1}{x^2}}
= x\left(1+\frac{1}{2x^2}-\frac{1}{8x^4}+\mathcal{O}\!\left(\frac{1}{x^6}\right)\right).
$$

Also

$$
\sinh^{-1}x=\ln\!\Big[x+\sqrt{1+x^2}\,\Big]
= \ln(2x)+\frac{1}{4x^2}-\frac{3}{32x^4}+\mathcal{O}\!\left(\frac{1}{x^6}\right).
$$

Start from

$$
f(x)=x(2x^2-3)\sqrt{1+x^2}+3\sinh^{-1}x
=(2x^2-3)\,x^2\left(1+\frac{1}{2x^2}-\frac{1}{8x^4}+\cdots\right)+3\sinh^{-1}x.
$$

### Expand the first piece

$$
(2x^2-3)\,x^2 = 2x^4 - 3x^2.
$$

Multiply term-by-term (keep down to $x^{-2}$):

$$
\begin{aligned}
(2x^4-3x^2)\left(1+\frac{1}{2x^2}-\frac{1}{8x^4}\right)
&= (2x^4-3x^2)
  + (2x^4-3x^2)\frac{1}{2x^2}
  - (2x^4-3x^2)\frac{1}{8x^4}
\\
&= \big[2x^4 - 3x^2\big]
 + \big[x^2 - \tfrac{3}{2}\big]
 + \big[-\tfrac{1}{4} + \tfrac{3}{8x^2}\big]
 + \mathcal{O}\!\left(\tfrac{1}{x^4}\right)
\\
&= 2x^4 - 2x^2 \;\;-\;\; \frac{7}{4} \;\;+\;\; \frac{1}{2x^2}
 + \mathcal{O}\!\left(\tfrac{1}{x^4}\right).
\end{aligned}
$$

### Add the second piece

$$
3\sinh^{-1}x
= 3\ln(2x) + \frac{3}{4x^2} - \frac{9}{32x^4} + \cdots
$$

### Combine

$$
\begin{aligned}
f(x)
&= \left[2x^4 - 2x^2 - \frac{7}{4} + \frac{1}{2x^2} + \cdots\right]
 + \left[3\ln(2x) + \frac{3}{4x^2} + \cdots\right]
\\[3pt]
&= \boxed{\,2x^4 - 2x^2\,} \;+\; 3\ln(2x) - \frac{7}{4}
   \;+\; \frac{5}{4x^2} + \cdots
\end{aligned}
$$

The problem statement only asks for the **leading polynomial terms**:

$$
\boxed{\,f(x)\approx 2x^4 - 2x^2 + \cdots\quad (x\to\infty)\,}.
$$


## Ultra-relativistic pressure from the series

Insert $f(x)\approx 2x^4$ into

$$
P_{e,r}=\frac{\pi m_0^4 c^5}{3h^3}\,f(x), \qquad x=\frac{p_0}{m_0 c},
$$

to obtain

$$
P_{e,r}\approx \frac{\pi m_0^4 c^5}{3h^3}\cdot 2\frac{p_0^4}{m_0^4 c^4}
= \boxed{\,\frac{2\pi c}{3h^3}\,p_0^4\,}.
$$

## Ultra-relativistic pressure from the $v_p=c$ integral

For a completely degenerate gas at $T=0$,

$$
P=\frac{1}{3\pi^2\hbar^3}\int_0^{p_0} v(p)\,p^3\,dp.
$$

In the **ultra-relativistic** limit $v(p)\simeq c$,

$$
P\simeq \frac{c}{3\pi^2\hbar^3}\int_0^{p_0} p^3\,dp
= \frac{c}{3\pi^2\hbar^3}\cdot \frac{p_0^4}{4}
= \frac{c}{12\pi^2\hbar^3}\,p_0^4.
$$

Using $\hbar = h/(2\pi)$,

$$
\frac{1}{12\pi^2\hbar^3}
= \frac{1}{12\pi^2}\cdot \frac{8\pi^3}{h^3}
= \frac{2\pi}{3h^3},
$$

hence

$$
\boxed{\,P=\frac{2\pi c}{3h^3}\,p_0^4\,}.
$$

This **matches exactly** the pressure from the large-$x$ limit of $f(x)$.


In [9]:
import math

def f_exact(x):
    return x*(2*x**2 - 3)*math.sqrt(1 + x**2) + 3*math.asinh(x)

def f_series_leading(x):
    # Leading two polynomial terms requested in the problem
    return 2*x**4 - 2*x**2

def f_series_with_logs(x):
    # Include the next subleading pieces we derived (useful to see residuals)
    return (2*x**4 - 2*x**2) + 3*math.log(2*x) - 7/4 + 5/(4*x**2)

print(f"{'x':>8} {'f_exact':>18} {'2x^4-2x^2':>18} {'rel.err (leading)':>18}  {'with logs':>18} {'rel.err (w/logs)':>18}")
for x in [3, 5, 10, 30, 100]:
    fe = f_exact(x)
    fl = f_series_leading(x)
    fr = f_series_with_logs(x)
    rel_lead = abs((fe - fl)/fe)
    rel_logs = abs((fe - fr)/fe)
    print(f"{x:8.1f} {fe:18.6e} {fl:18.6e} {rel_lead:18.3e}  {fr:18.6e} {rel_logs:18.3e}")


       x            f_exact          2x^4-2x^2  rel.err (leading)           with logs   rel.err (w/logs)
     3.0       1.477578e+02       1.440000e+02          2.543e-02        1.477642e+02          4.286e-05
     5.0       1.205207e+03       1.200000e+03          4.320e-03        1.205208e+03          7.091e-07
    10.0       1.980725e+04       1.980000e+04          3.660e-04        1.980725e+04          2.745e-09
    30.0       1.618211e+06       1.618200e+06          6.510e-06        1.618211e+06          4.170e-13
   100.0       1.999800e+08       1.999800e+08          7.073e-08        1.999800e+08          1.490e-16


## Final Results (boxed)

Large-$x$ asymptotics:
$$
\boxed{\,f(x)\approx 2x^4 - 2x^2 + \cdots \quad (x\to\infty)\,}
$$

Ultra-relativistic degenerate pressure:
$$
\boxed{\,P_{e,r}\approx \frac{2\pi c}{3h^3}\,p_0^4\,}
$$

Pressure from the $v_p=c$ integral:
$$
\boxed{\,P=\frac{2\pi c}{3h^3}\,p_0^4\,}
$$

These two expressions are **identical**, confirming that the $x\to\infty$ limit of the
fully relativistic formula reduces to the standard ultra-relativistic (massless-speed) result.


## Discussion

For large $x$, the function $f(x)$ grows roughly as $2x^4$, showing the **ultra-relativistic** scaling where $P \propto p_0^4$.  
The next correction, $-2x^2$, slightly reduces the pressure at lower $x$.  

Substituting $f(x)\!\sim\!2x^4$ into the expression for $P_{e,r}$ gives the same form obtained by setting $v_p = c$ in the pressure integral, confirming the consistency of the limit.  

Numerically, this two-term approximation reproduces $f(x)$ very well for $x \gtrsim 10$, with only minor differences at moderate $x$.



## Problem 6 — Direct derivation of the doubled factor

Start from the standard formulas (non-relativistic, $T=0$ degeneracy vs. Maxwellian):
$$
P_{e,\mathrm{deg}}=\frac{1}{5}\frac{\hbar^2}{m_e}(3\pi^2)^{2/3}n_e^{5/3}
\equiv A\,n_e^{5/3},\qquad
P_{e,\mathrm{MB}}=n_e k_B T.
$$

Form the ratio:
$$
\frac{P_{e,\mathrm{deg}}}{P_{e,\mathrm{MB}}}
=\frac{A}{k_B}\,\frac{n_e^{2/3}}{T}.
$$

Relate $n_e$ to density:
$$
n_e=\frac{\rho}{\mu_e m_u}.
$$

Now **plug in the locus given in the problem**,
$$
\frac{\rho}{\mu_e}=C_0\,T^{3/2},\qquad C_0=5\times10^{-8}\ \mathrm{g\,cm^{-3}\,K^{-3/2}},
$$
which gives
$$
n_e=\frac{C_0\,T^{3/2}}{m_u}
\quad\Longrightarrow\quad
n_e^{2/3}=\left(\frac{C_0}{m_u}\right)^{2/3}T.
$$

Therefore the ratio becomes a **pure number** (no $T$ dependence):
$$
\frac{P_{e,\mathrm{deg}}}{P_{e,\mathrm{MB}}}
=\frac{A}{k_B}\left(\frac{C_0}{m_u}\right)^{2/3},
\qquad
A=\frac{1}{5}\frac{\hbar^2}{m_e}(3\pi^2)^{2/3}.
$$

Using the usual classroom cgs constants 
($\hbar=1.055\times10^{-27}$ erg·s, $m_e=9.11\times10^{-28}$ g, 
$k_B=1.381\times10^{-16}$ erg·K$^{-1}$, $m_u=1.661\times10^{-24}$ g) and 
$(3\pi^2)^{2/3}\simeq 9.61$, we find numerically
$$
\frac{P_{e,\mathrm{deg}}}{P_{e,\mathrm{MB}}}
\approx \underbrace{\frac{1}{5}\frac{\hbar^2}{m_e}\frac{(3\pi^2)^{2/3}}{k_B}}_{\displaystyle \approx\,1.70\times10^{-11}}
\;\times\;
\underbrace{\left(\frac{5\times10^{-8}}{1.661\times10^{-24}}\right)^{2/3}}_{\displaystyle \approx\,9.68\times10^{10}}
\;\approx\;\boxed{\,1.65\,}.
$$

So, **on the problem’s quoted curve** $\rho/\mu_e=5\times10^{-8}T^{3/2}$,
$$
\boxed{\;\frac{P_{e,\mathrm{deg}}}{P_{e,\mathrm{MB}}}\ \approx\ 1.65\ \ \text{(i.e., “of order twice”)}\;}
$$
with standard rounded constants used in class.


## Problem 7: Degeneracy, relativity, and electron pressure for a CO mixture

A gas composed of ${}^{12}\mathrm{C}$ and ${}^{16}\mathrm{O}$ has density $\rho=2.5\times 10^{5}\ \mathrm{g\,cm^{-3}}$ at $T=10^{8}\ \mathrm{K}$.

**Tasks**  
1) Determine whether the gas is degenerate or non-degenerate (in the $(\log T,\log\rho)$ plane sense).  
2) If completely degenerate, decide whether it is non-relativistic, partially relativistic, or extremely relativistic.  
3) Compute $P_e$ from the **complete-degenerate** relativistic table **$f(x)$ vs $x$**.  
4) Assuming **incomplete, non-relativistic** degeneracy, compute $P_e$ using the **Fermi–Dirac functions** table $F_{1/2}(\eta), F_{3/2}(\eta)$.  
5) Compare the pressures and comment: which is more correct here and why?  
6) What is the ratio of **electron** to **ion** pressure?

**Assumptions:** 50/50 **by mass** C/O. Then $Z/A=1/2$ for both species, so

$$
\mu_e = 2,\qquad 
\frac{1}{\mu_i}=\frac{X_C}{A_C}+\frac{X_O}{A_O}
=\frac{1}{2}\left(\frac{1}{12}+\frac{1}{16}\right)
=\frac{7}{96}
\;\Rightarrow\;
\mu_i=\frac{96}{7}\approx 13.7143.
$$


## Derivation

**Electron number density**

$$
n_e=\frac{\rho}{\mu_e m_u}=\frac{2.5\times 10^{5}}{2\,m_u}\ \ \mathrm{cm^{-3}}.
$$

**Fermi momentum and relativity parameter**

$$
p_F=\hbar\,(3\pi^2 n_e)^{1/3},\qquad x\equiv\frac{p_F}{m_e c}.
$$

- Relativistic regime classifier: $x\ll1$ (NR), $x\sim1$ (partially relativistic), $x\gg1$ (UR).

**Degeneracy criterion** (compare Fermi energy to thermal energy):

$$
E_F^{(\mathrm{kin})}=\sqrt{(p_F c)^2+(m_e c^2)^2}-m_e c^2,\qquad \text{degenerate if } E_F^{(\mathrm{kin})}\gg k_B T.
$$


### Complete degeneracy (use $f(x)$ table)

Given the tabulated function $f(x)$, the **zero-temperature** relativistic electron pressure is

$$
P_{e,\mathrm{deg}}=\frac{\pi m_e^4 c^5}{3h^3}\,f(x).
$$

We obtain $x$ from $n_e$, interpolate $f(x)$ from the **CompleteDegenerateRelativistic** table, then evaluate $P_{e,\mathrm{deg}}$.


### Incomplete, non-relativistic degeneracy (use FD table)

For **finite $T$**, non-relativistic electrons:

$$
n_e = \frac{2}{\sqrt{\pi}\,\lambda_T^3}\,F_{1/2}(\eta),\qquad
P_e = \frac{2 k_B T}{\sqrt{\pi}\,\lambda_T^3}\,F_{3/2}(\eta),
$$

with thermal de Broglie wavelength

$$
\lambda_T = \frac{h}{\sqrt{2\pi m_e k_B T}}.
$$

Procedure: solve the first equation for $\eta$ (interpolate $F_{1/2}$ from the **FermiDiracFunctions** table), then plug $\eta$ into the second (interpolate $F_{3/2}$).


### Ion pressure

Treat ions as an ideal classical gas:

$$
n_i=\frac{\rho}{\mu_i m_u},\qquad P_i=n_i k_B T.
$$


In [12]:
# --- Problem 7 (table-driven): degeneracy, relativity, and pressures ---
import numpy as np, math, pandas as pd
from pathlib import Path

# ---------- Constants (cgs) ----------
h   = 6.62607015e-27
hbar= h/(2*math.pi)
kB  = 1.380649e-16
m_u = 1.66053906660e-24
m_e = 9.1093837015e-28
c   = 2.99792458e10

# ---------- Given ----------
rho = 2.5e5   # g/cm^3
T   = 1.0e8   # K

# 50/50 by mass C/O
mu_e = 2.0
mu_i = 96.0/7.0

# ---------- Electron quantities ----------
n_e = rho/(mu_e*m_u)
pF  = hbar * (3.0*math.pi**2 * n_e)**(1.0/3.0)
x   = pF/(m_e*c)
EF_tot = math.sqrt((pF*c)**2 + (m_e*c**2)**2)   # total Fermi energy (erg)
EF_kin = EF_tot - m_e*c**2                      # kinetic part (erg)
deg_ratio = EF_kin/(kB*T)

# ---------- Load tables ----------
path_rel = Path("CompleteDegenerateRelativistic.txt")
path_fd  = Path("FermiDiracFunctions.txt")

if not path_rel.exists():
    raise FileNotFoundError(f"Missing {path_rel.name}. Put it next to the notebook.")
if not path_fd.exists():
    raise FileNotFoundError(f"Missing {path_fd.name}. Put it next to the notebook.")

# 1) Relativistic degenerate f(x) table: columns 'x  f(x)'
rel_rows = []
for line in path_rel.read_text(errors="ignore").splitlines():
    s=line.strip()
    if (not s) or s.startswith("#"): 
        continue
    parts = s.split()
    if len(parts)>=2:
        # accept scientific notation with E or e
        xv = float(parts[0].replace("D","E").replace("d","e"))
        fv = float(parts[1].replace("D","E").replace("d","e"))
        rel_rows.append([xv, fv])

rel_df = pd.DataFrame(rel_rows, columns=["x","f"]).dropna()
rel_df = rel_df.sort_values("x")
x_tab  = rel_df["x"].to_numpy()
f_tab  = rel_df["f"].to_numpy()

# clamp x into table range, then interpolate f(x)
if x < x_tab[0] or x > x_tab[-1]:
    # outside table: use nearest value (instructor-provided tables often cover 1e-3..1e2)
    x_clip = min(max(x, x_tab[0]), x_tab[-1])
else:
    x_clip = x
f_x = np.interp(x_clip, x_tab, f_tab)

# Degenerate electron pressure from table f(x)
P_e_deg = (math.pi * m_e**4 * c**5)/(3.0*h**3) * f_x  # dyn/cm^2

# 2) FD table: columns '# alpha   (2/3)*F_3/2    F_1/2'
fd_rows = []
for line in path_fd.read_text(errors="ignore").splitlines():
    s=line.strip()
    if (not s) or s.startswith("#"):
        continue
    p=s.split()
    if len(p)>=3:
        a    = float(p[0].replace("D","E").replace("d","e"))
        two3 = float(p[1].replace("D","E").replace("d","e"))
        F12  = float(p[2].replace("D","E").replace("d","e"))
        fd_rows.append([a, two3, F12])

fd = pd.DataFrame(fd_rows, columns=["alpha","two_thirds_F32","F12"]).dropna()
# sort by alpha, and also build monotone arrays by F12 for inversion
fd = fd.sort_values("alpha")
alpha_arr = fd["alpha"].to_numpy()
F12_arr   = fd["F12"].to_numpy()
two3_arr  = fd["two_thirds_F32"].to_numpy()

# Therm. wavelength and target F_{1/2}
lambda_T = h/ math.sqrt(2.0*math.pi*m_e*kB*T)
F12_target = n_e * (math.sqrt(math.pi)/2.0) * (lambda_T**3)

# Ensure monotonicity (typical tables are increasing in alpha). If not, sort by F12.
if not np.all(np.diff(F12_arr) > 0):
    order = np.argsort(F12_arr)
    F12_arr   = F12_arr[order]
    alpha_arr = alpha_arr[order]
    two3_arr  = two3_arr[order]

# Bound the target to table range to avoid NaNs, then invert by linear interpolation:
F12_min, F12_max = F12_arr[0], F12_arr[-1]
if F12_target <= F12_min:
    alpha_star = alpha_arr[0]
elif F12_target >= F12_max:
    alpha_star = alpha_arr[-1]
else:
    alpha_star = np.interp(F12_target, F12_arr, alpha_arr)

# Now get (2/3)F_{3/2} at alpha_star by interpolation on alpha
two3_vs_alpha = two3_arr
# If we earlier reordered by F12, alpha_arr is aligned; we can safely interpolate in alpha
two3_star = np.interp(alpha_star, alpha_arr, two3_vs_alpha)
F32_star  = (3.0/2.0) * two3_star

# Finite-T (incomplete) nonrelativistic electron pressure from FD table
P_e_NRinc = (2.0*kB*T/ math.sqrt(math.pi)) * (1.0/(lambda_T**3)) * F32_star

# Ion pressure
n_i = rho/(mu_i*m_u)
P_i = n_i*kB*T

# Classifications
rel_regime = "non-relativistic" if x<0.1 else ("partially relativistic" if x<3.0 else "ultra-relativistic")
deg_regime = "strongly degenerate" if deg_ratio>5 else ("moderately degenerate" if deg_ratio>1 else "non-degenerate")

# ---------- Output ----------
print("=== Problem 7: Table-driven results ===")
print(f"mu_e={mu_e:.3f}, mu_i={mu_i:.6f}")
print(f"n_e = {n_e:.6e} cm^-3")
print(f"p_F = {pF:.6e} g cm/s,   x = p_F/(m_e c) = {x:.6f}  --> {rel_regime}")
print(f"E_F(kin)/kT = {deg_ratio:.3f}  --> {deg_regime}")
print(f"From table: f(x={x_clip:.6f}) = {f_x:.6e}")
print(f"P_e (degenerate; table f) = {P_e_deg:.6e} dyn/cm^2")
print(f"alpha* = {alpha_star:.6f},   F12_target = {F12_target:.6e}")
print(f"(2/3)F_3/2(alpha*) = {two3_star:.6e}  =>  F_3/2(alpha*) = {F32_star:.6e}")
print(f"P_e (incomplete NR; FD table) = {P_e_NRinc:.6e} dyn/cm^2")
print(f"P_i (ideal ions) = {P_i:.6e} dyn/cm^2")
print(f"Ratios:  P_e(deg)/P_i = {P_e_deg/P_i:.3f},   P_e(NRinc)/P_i = {P_e_NRinc/P_i:.3f},   P_e(deg)/P_e(NRinc) = {P_e_deg/P_e_NRinc:.3f}")


=== Problem 7: Table-driven results ===
mu_e=2.000, mu_i=13.714286
n_e = 7.527676e+28 cm^-3
p_F = 1.377538e-17 g cm/s,   x = p_F/(m_e c) = 0.504422  --> partially relativistic
E_F(kin)/kT = 7.117  --> strongly degenerate
From table: f(x=0.504422) = 4.896875e-02
P_e (degenerate; table f) = 2.939267e+21 dyn/cm^2
alpha* = -11.905990,   F12_target = 2.762776e+01
(2/3)F_3/2(alpha*) = 2.796389e+02  =>  F_3/2(alpha*) = 4.194583e+02
P_e (incomplete NR; FD table) = 1.577928e+22 dyn/cm^2
P_i (ideal ions) = 1.515657e+20 dyn/cm^2
Ratios:  P_e(deg)/P_i = 19.393,   P_e(NRinc)/P_i = 104.109,   P_e(deg)/P_e(NRinc) = 0.186


## Final Results (boxed)

**Composition assumption:** 50/50 by mass CO mixture ⇒ $\mu_e = 2.000$, $\mu_i \approx 13.714286$.


### Basics
$$
\boxed{\,n_e = 7.527676\times 10^{28}\ \mathrm{cm^{-3}}\,}
\qquad
\boxed{\,p_F = 1.377538\times 10^{-17}\ \mathrm{g\,cm\,s^{-1}}\,}
$$


### Classification
$$
\boxed{\,x=\frac{p_F}{m_e c}=0.504422\ \Rightarrow\ \text{partially relativistic}\,}
\qquad
\boxed{\,\frac{E_F^{(\mathrm{kin})}}{k_B T}=7.117\ \Rightarrow\ \text{strongly degenerate}\,}
$$


### Complete-degenerate (table \(f(x)\))
$$
\boxed{\,f(x=0.504422)=4.896875\times 10^{-2}\,}
\qquad
\boxed{\,P_{e,\mathrm{deg}}=2.939267\times 10^{21}\ \mathrm{dyn\,cm^{-2}}\,}
$$


### Incomplete, non-relativistic (FD table)
$$
\boxed{\,\alpha^\ast=\eta=-11.905990,\quad F_{1/2}(\eta)=2.762776\times 10^{1},\quad F_{3/2}(\eta)=4.194583\times 10^{2}\,}
$$

$$
\boxed{\,P_{e,\mathrm{NR,inc}}=1.577928\times 10^{22}\ \mathrm{dyn\,cm^{-2}}\,}
$$


### Ions (ideal)
$$
\boxed{\,P_i=1.515657\times 10^{20}\ \mathrm{dyn\,cm^{-2}}\,}
$$


### Ratios
$$
\boxed{\,\frac{P_{e,\mathrm{deg}}}{P_i}=19.393,\quad
\frac{P_{e,\mathrm{NR,inc}}}{P_i}=104.109,\quad
\frac{P_{e,\mathrm{deg}}}{P_{e,\mathrm{NR,inc}}}=0.186\,}
$$


## Discussion

At $\rho = 2.5\times10^5\ \mathrm{g\,cm^{-3}}$ and $T = 10^8\ \mathrm{K}$, the electrons have 
$x \approx 0.50$ and $E_F^{(\mathrm{kin})}/kT \approx 7.1$, meaning the gas is **strongly degenerate** and **partially relativistic**.  
This places the CO mixture well within the degenerate region of the EOS diagram.

The **degenerate relativistic** pressure from the $f(x)$ table,
$$
P_{e,\mathrm{deg}} = 2.94\times10^{21}\ \mathrm{dyn\,cm^{-2}},
$$
is smaller than the **finite-$T$ non-relativistic** value,
$$
P_{e,\mathrm{NR,inc}} = 1.58\times10^{22}\ \mathrm{dyn\,cm^{-2}},
$$
because relativistic effects slightly soften the EOS.  
A small finite-temperature correction ($\sim1\%$) would bring the $T=0$ result even closer to the true value.

The **ion pressure** is much smaller,
$$
P_i = 1.52\times10^{20}\ \mathrm{dyn\,cm^{-2}},
$$
so electrons clearly dominate the total pressure.  
Overall, the system is best described by the **degenerate relativistic** case, with only minor corrections from temperature or ion contributions.
